src/tpm2_param_enc.c

```diff
@@ -0,0 +1,306 @@
+/* --- Local Functions -- */
+/******************************************************************************/
+
+/* This function performs key generation according to Part 1 of the TPM spec
+ * and returns the number of bytes generated, which may be zero.
+ *
+ * 'key' and 'keyStream' are pointers and they can not be NULL.
+ * Any other pointer arguments may be NULL.
+ * 'sizeInBits' must be no larger than (2^18)-1 = 256K bits (32385 bytes).
+ *
+ * Note: The "once" parameter is set to allow incremental generation of a large
+ * value. If this flag is TRUE, "sizeInBits" is used in the HMAC computation
+ * but only one iteration of the KDF is performed. This would be used for
+ * XOR obfuscation so that the mask value can be generated in digest-sized
+ * chunks rather than having to be generated all at once in an arbitrarily
+ * large buffer and then XORed into the result. If "once" is TRUE, then
+ * "sizeInBits" must be a multiple of 8.
+ *
+ * Any error in the processing of this command is considered fatal.
+ *
+ * Return values:
+ *     0    hash algorithm is not supported or is TPM_ALG_NULL
+ *    >0    the number of bytes in the 'keyStream' buffer
+ *
+ */
+static int TPM2_KDFa(
+    TPM_ALG_ID  hashAlg,    /* IN: hash algorithm used in HMAC */
+    TPM2B_DATA  *key,       /* IN: key, can be NULL if authValue is used */
+    const char  *label,     /* IN: a 0-byte terminated label used in KDF */
+    TPM2B_DATA  *contextU,  /* IN: context U */
+    TPM2B_DATA  *contextV,  /* IN: context V */
+    UINT32      sizeInBits, /* IN: size of generated key in bits */
+    BYTE        *keyStream, /* OUT: key buffer */
+    UINT32      *counterInc,/* IN/OUT: See Note for incremental operations */
+    int         doOnce      /* IN: TRUE if only one iteration is performed */
+)
+{
+    int ret, hashType;
+    Hmac hmac;
+    word32 counter = 0;
+    int hLen, outLen, lLen = 0;
+    byte *outStream;
+    byte uint32Buf[sizeof(UINT32)];
+
+    if (key == NULL || keyStream == NULL)
+        return BAD_FUNC_ARG;
+
+    if (doOnce != 0 && (sizeInBits & 7) != 0)
+        return BAD_FUNC_ARG;
+
+    hashType = TPM2_GetHashType(hashAlg);
+    if (hashType == WC_HASH_TYPE_NONE)
+        return NOT_COMPILED_IN;
+
+    hLen = TPM2_GetHashDigestSize(hashAlg);
+    if (hLen <= 0)
+        return NOT_COMPILED_IN;
+
+    ret = wc_HmacInit(&hmac, NULL, INVALID_DEVID);
+    if (ret != 0)
+        return ret;
+
+    /* setup counter */
+    counter = (counterInc != NULL) ? *counterInc : 0;
+
+    /* get label length if provided */
+    if (label != NULL)
+        lLen = (int)XSTRLEN(label);
+
+    /* generate required bytes */
+    outLen = doOnce ? hLen : (((int)sizeInBits + 7) / 8);
+    outStream = keyStream;
+
+    for (; outLen > 0; outLen -= hLen)
+    {
+        if (hLen > outLen)
+            hLen = outLen;
+
+        counter++;
+
+        /* start HMAC */
+        ret = wc_HmacSetKey(&hmac, hashType, &key->buffer[0], key->size);
```

Comment:
```
src/tpm2_param_enc.c:128:15: warning: Function call argument is an uninitialized value
        ret = wc_HmacSetKey(&hmac, hashType, &key->buffer[0], key->size);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
please check if you get this with after the latest commits